kof-operator/internal/controller/utils/utils.go

```diff
@@ -0,0 +1,90 @@
+package utils
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+
+	kcmv1alpha1 "github.com/K0rdent/kcm/api/v1alpha1"
+	"github.com/k0rdent/kof/kof-operator/internal/controller/record"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const ManagedByLabel = "app.kubernetes.io/managed-by"
+const ManagedByValue = "kof-operator"
+
+func GetOwnerReference(owner client.Object, client client.Client) (metav1.OwnerReference, error) {
+	gvk := owner.GetObjectKind().GroupVersionKind()
+
+	if gvk.Empty() {
+		var err error
+		gvk, err = client.GroupVersionKindFor(owner)
+		if err != nil {
+			return metav1.OwnerReference{}, err
+		}
+	}
+
+	return metav1.OwnerReference{
+		APIVersion: gvk.GroupVersion().String(),
+		Kind:       gvk.Kind,
+		Name:       owner.GetName(),
+		UID:        owner.GetUID(),
+	}, nil
+}
+
+func BoolPtr(value bool) *bool {
+	// `*bool` fields may point to `true`, `false`, or be `nil`.
+	// Direct `&true` is an error.
+	return &value
+}
+
+func GetEventsAnnotations(cd *kcmv1alpha1.ClusterDeployment) map[string]string {
+	var generation string
+
+	if cd.Generation == 0 {
+		generation = "nil"
+	} else {
+		generation = strconv.Itoa(int(cd.Generation))
+	}
+
+	return map[string]string{
+		"generation": generation,
+	}
+}
+
+func GetClusterDeploymentStub(name, namespace string) *kcmv1alpha1.ClusterDeployment {
+	return &kcmv1alpha1.ClusterDeployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "k0rdent.mirantis.com/v1alpha1",
+			Kind:       kcmv1alpha1.ClusterDeploymentKind,
+		},
+	}
+}
+
+func HandleError(ctx context.Context, reason, message string, cd *kcmv1alpha1.ClusterDeployment, err error, logKeysAndValues ...any) {
+	log := log.FromContext(ctx)
+	log.Error(err, message, logKeysAndValues...)
+
+	formattedKeysValues := make([]string, 0, len(logKeysAndValues))
+	for i, value := range logKeysAndValues {
+		if i != 0 && i%2 == 0 {
+			formattedKeysValues = append(formattedKeysValues, fmt.Sprintf("%v, ", value))
+		} else {
+			formattedKeysValues = append(formattedKeysValues, fmt.Sprintf("%v ", value))
+		}
+	}
```

Comment:
You already allocated an array with len, so you don't need to append, otherwise you get an empty slots at the beginning of the array.
```suggestion
	formattedKeysValues := make([]string, 0, len(logKeysAndValues))
	for i, value := range logKeysAndValues {
            formattedKeysValues[i] = fmt.Sprintf("%v, ", value)
	}
```

Reply:
I guess the intention was to convert e.g. `["description", "key1", "value1", "key2", "value2"]` to `description key1 value1, key2 value2`. Please try to get `description, key1: value1, key2: value2` or `description, key1=value1, key2=value2`.

Reply:
In Go append() utilizes pre-allocated memory until capacity is exhausted.
[Link to the article](https://medium.com/@ksandeeptech07/zero-allocation-in-go-ce29e6a9ffdc)